### PR TITLE
Add percentile capabilities to sqlserver

### LINKED ISSUE #12222: Custom expression function Median() fails against SQL Server
**Edit:** Just finished reading the docs for expressions and noted the limitations section listing these functions. Going to leave it open though, please consider it a feature request :)

**Describe the bug**
Custom expression Median() is not working against SQL Server data sources. Get an error in the notebook editor when previewed.

I'm so looking forward to this working, we use quite a few median's in our reporting.

**Logs**
Nothing in either the javascript console or the server logs.

**To Reproduce**
1. Start a new question against a data source from SQL Server
2. Summarise using a Custom Expression
3. Expression "Median([number field])", give it a name
4. Hit the preview button
5. Error displayed in the preview pane: "The function 'PERCENTILE_CONT' must have an OVER clause."

**Expected behavior**
The median is calculated and presented to the user

**Screenshots**
image

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "Cp1252",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "13.0.2+8",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "https://java.oracle.com/",
    "java.version": "13.0.2",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "13.0.2+8",
    "os.name": "Windows 10",
    "os.version": "10.0",
    "user.language": "en",
    "user.timezone": "Pacific/Auckland"
  },
  "metabase-info": {
    "databases": [
      "sqlserver",
      "postgres",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "12.1"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-03-31",
      "tag": "v0.35.0",
      "branch": "master",
      "hash": "3f96c47"
    },
    "settings": {
      "report-timezone": "Pacific/Auckland"
    }
  }
}
```

**Severity**
This is a new function so doesn't break anything for us but it was one we were very excited to see coming, shame it's not working for us. We can continue to calculate medians outside of Metabase but it's a frustration.

**Additional context**
The Median() function seems to work fine against the Sample Dataset, it seems to be a SQL Server syntax issue.

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes #12222

### Description

**This PR should not be merged before CI is modified to contain sqlserver 2022 job.**

This PR takes advantage of the fact, that sqlserver 2022 adds [`APPROX_PERCENTILE_CONT`](https://learn.microsoft.com/en-us/sql/t-sql/functions/approx-percentile-disc-transact-sql?view=sql-server-ver16) aggregate function and enables `:percentile-aggregations` for the versions 2022 (16.0) and above.

### How to verify

Run the `metabase.query-processor-test.advanced-math-test/test-percentile` or try it out using query builder.

### CODE REVIEW

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj [hunk 1]
```diff
@@ -472,6 +488,14 @@
   [driver [_ arg power]]
   (hx/call :power (hx/cast :float (sql.qp/->honeysql driver arg)) (sql.qp/->honeysql driver power)))
 
+(defmethod sql.qp/->honeysql [:sqlserver :percentile]
+  [driver [_ arg val]]
+  (hx/raw (apply format "APPROX_PERCENTILE_CONT (%s) WITHIN GROUP (ORDER BY %s)"
```
Comment (lbrdnk) [inline modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj:493]: I believe this does not open us to any vulnerability thanks to use of `sql.qp/->honeysql`. Let me know if that's just naive and I'll further investigate.

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj [hunk 2]
```diff
@@ -32,14 +33,29 @@
...
                               :convert-timezone                       true
                               :test/jvm-timezone-setting              false}]
   (defmethod driver/database-supports? [:sqlserver feature] [_driver _feature _db] supported?))
 
+(defn- dbms-major-version
+  [db]
+  ;;;; Using try-catch as either destructuring can throw, or returned could be a non number. Happened previously with
+  ;;;; mongo. See issue #29678.
+  (try
+    (let [{[major-version] :semantic-version} (driver/dbms-version :sqlserver db)]
```
Comment (metamben) [inline modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj:48]: This is expensive. Have a look at how the Mongo driver does this checks.

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj (hunk 1 above)
Comment (metamben) [reply modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj:493]: I don't expect `sql.qp/->honeysql` would prevent injections. (Using prepared statement parameters for input does for sure.) Tests are needed anyway, so add some and try to inject something to see if you can convince yourself and others that this is not a problem.

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj (hunk 2 above)
Comment (lbrdnk) [reply modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj]: I've modified the code to retrieve `:dbms_version` right from the `db` structure.

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj (hunk 1 above)
Comment (lbrdnk) [reply modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj]: I've changed the code to use `hformat/fn-handler` to extend honeysql with `::approx-percentile-cont` function. With regards to injections, I've tried various ways how to camouflage some code into parameters, unsuccessfully.

**Review by metamben — APPROVED** [review]
It's a bit sad that this is still using honeysql 1, but LGTM.

**Pull request merged**